### PR TITLE
Added secure_boot and affinity commands

### PR DESCRIPTION
**Pull request by Boredphilosopher96**

### CODE REVIEW

File: vhpc_toolkit/get_args.py [hunk 1]
```diff
@@ -246,6 +246,135 @@ def get_args():
     )
     power_group2.add_argument("--on", action="store_true", help="Power on")
     power_group2.add_argument("--off", action="store_true", help="Power off")
+
+    secure_boot_parser = subparsers.add_parser(
+        "secure_boot",
+        help="Secure boot on/off VM(s)",
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Change help msg to "Turn secure boot on/off for VM(s)"

File: vhpc_toolkit/get_args.py [hunk 2]
```diff
@@ -246,6 +246,135 @@ def get_args():
...
+    numa_affinity_group2 = numa_affinity_parser.add_mutually_exclusive_group(
+        required=True
+    )
+
+    numa_affinity_group1.add_argument(
+        "--vm",
+        action="store",
+        default=None,
+        type=str,
+        help="Name of the VM on which to NUMA node affinity",
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Name of the VM on which to *change* NUMA node affinity

File: vhpc_toolkit/get_args.py [hunk 3]
```diff
@@ -246,6 +246,135 @@ def get_args():
...
+                )
+        return affinity_list
+
+    vm_affinity_parser = subparsers.add_parser(
+        "vm_scheduling_affinity",
+        help="Change VM scheduling affinity",
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    vm_affinity_group1 = vm_affinity_parser.add_mutually_exclusive_group(required=True)
+    vm_affinity_group2 = vm_affinity_parser.add_mutually_exclusive_group(required=True)
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py:325]: This doesn't need to be a mutually exclusive group

File: vhpc_toolkit/get_args.py [hunk 4]
```diff
@@ -246,6 +246,135 @@ def get_args():
...
+
+    numa_affinity_parser = subparsers.add_parser(
+        "numa_affinity",
+        help="Change NUMA node affinity",
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    numa_affinity_group1 = numa_affinity_parser.add_mutually_exclusive_group(
+        required=True
+    )
+    numa_affinity_group2 = numa_affinity_parser.add_mutually_exclusive_group(
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py:365]: This doesn't need to be a mutually exclusive group

File: README.md [hunk 5]
```diff
@@ -93,7 +93,12 @@ cd vhpc-toolkit
 virtualenv -p python3 venv
 source venv/bin/activate
 pip install -r requirements.txt
-python setup.py install
+pip install vhpc_toolkit
```
Comment (Michael870) [inline README.md:96]: This doesn't work for me. Error msg below: 

(venv) xiaolongc-a02:vhpc-toolkit xiaolongc$ pip install -e vhpc_toolkit
Obtaining file:///Users/xiaolongc/Documents/vhpc_toolkit/github/vhpc-toolkit/vhpc_toolkit
ERROR: file:///Users/xiaolongc/Documents/vhpc_toolkit/github/vhpc-toolkit/vhpc_toolkit does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.

File: vhpc_toolkit/get_args.py [hunk 6]
```diff
@@ -511,6 +640,21 @@ def get_args():
         help="Name of the file containing a list of VMs, "
         "one per line, to perform the passthrough operation",
     )
+    passthru_group1.add_argument(
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Can we separate the code for passthru_host from this merge request?

File: vhpc_toolkit/operations.py [hunk 7]
```diff
@@ -668,6 +711,17 @@ def network_cli(self):
             if tasks:
                 GetWait().wait_for_tasks(tasks, task_name="Remove network adapter(s)")
 
+    def change_affinity_cli(self, vm_name, affinity):
```
Comment (Michael870) [inline vhpc_toolkit/operations.py]: is this function obsolete?

File: vhpc_toolkit/get_args.py (hunk 3 above)
Comment (Michael870) [reply vhpc_toolkit/get_args.py:325]: Actually, can we add another argument like "--clear" to clear the affinity?

File: vhpc_toolkit/get_args.py (hunk 4 above)
Comment (Michael870) [reply vhpc_toolkit/get_args.py:365]: Same as above, add another argument --clear

File: vhpc_toolkit/command.py [hunk 8]
```diff
@@ -59,3 +59,9 @@ def command():
         ops.cluster()
     elif ops.cfg[CMD_KEY] == "get_vm_config":
         ops.get_vm_config_cli()
+    elif ops.cfg[CMD_KEY] == "secure_boot":
+        ops.secure_boot_cli()
+    elif ops.cfg[CMD_KEY] == "vm_scheduling_affinity":
```
Comment (Michael870) [inline vhpc_toolkit/command.py]: Let's shorten this command to sched_affinity. The current one is too long and mess up with the help msg:

get_vm_config       View the performance metrics of the VM
    view                View the vCenter object names
    clone               Clone VM(s) via Full Clone or Linked Clone
    destroy             Destroy VM(s)
    power               Power on/off VM(s)
    secure_boot         Secure boot on/off VM(s)
    vm_scheduling_affinity
                        Change VM scheduling affinity
    numa_affinity       Change NUMA node affinity
    cpumem              Reconfigure CPU/memory for VM(s)

**Review by Michael870 — CHANGES_REQUESTED** [review]

File: vhpc_toolkit/operations.py [hunk 9]
```diff
@@ -647,6 +648,48 @@ def _get_poweroff_tasks(self, vms):
...
+        Returns:
+            List of task objects
+
+        """
+        tasks = []
+        for vm in vms:
+            vm_obj = self.objs.get_vm(vm)
+            if GetVM(vm_obj).is_power_on():
+                self.logger.info(
+                    "VM {0} is turned on. So cannot enable secure boot. Please turn off and try again".format(
```
Comment (Michael870) [inline vhpc_toolkit/operations.py]: "cannot enable secure boot" -> "cannot change secure boot"

File: vhpc_toolkit/get_args.py [hunk 10]
```diff
@@ -246,6 +246,156 @@ def get_args():
...
+        help="Name of the file containing a list of VMs, one per line,"
+        " to reconfigure vm scheduling affinity",
+    )
+
+    vm_affinity_group2.add_argument(
+        "--affinity",
+        action="store",
+        default=[],
+        type=affinity_array,
+        help="Use '-' for ranges and ',' to separate values."
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: Use : for ranges

File: vhpc_toolkit/get_args.py [hunk 11]
```diff
@@ -246,6 +246,156 @@ def get_args():
...
+        help="Name of the file containing a list of VMs, one per line,"
+        " to reconfigure NUMA node affinity",
+    )
+
+    numa_affinity_group2.add_argument(
+        "--affinity",
+        action="store",
+        default=[],
+        type=affinity_array,
+        help="Use '-' for ranges and ',' to separate values."
```
Comment (Michael870) [inline vhpc_toolkit/get_args.py]: same as above

File: vhpc_toolkit/operations.py [hunk 12]
```diff
@@ -2095,3 +2138,42 @@ def _print_vm_config(self, vm_name: str) -> None:
...
+                    )
+                )
+            else:
+                tasks.append(
+                    ConfigVM(vm_obj).change_vm_scheduling_affinity(self.cfg["affinity"])
+                )
+
+        GetWait().wait_for_tasks(tasks, task_name="Set VM scheduling affinity")
+
+    def numa_affinity_cli(self):
```
Comment (Michael870) [inline vhpc_toolkit/operations.py:2166]: I think you also need to check for VM power on/off before the change

**Review by Michael870 — CHANGES_REQUESTED** [review]
Good job. At this point only a few minor changes are needed before you can merge.

**Review by Michael870 — APPROVED** [review]

**Review by Boredphilosopher96 — COMMENTED** [review]
Merging

**Pull request merged**